GraknClient.java

```diff
@@ -19,63 +19,159 @@
+import static grakn.common.collection.Collections.pair;
 
-    public static final String DEFAULT_URI = "localhost:1729";
+public class GraknClient {
+    public static final String DEFAULT_ADDRESS = "localhost:1729";
 
-    private final ManagedChannel channel;
-    private final DatabaseManager databases;
+    public static class Core implements Grakn.Client {
+        private final ManagedChannel channel;
+        private final RPCDatabaseManager.Core databases;
 
-    public GraknClient() {
-        this(DEFAULT_URI);
-    }
+        public Core() {
+            this(DEFAULT_ADDRESS);
+        }
 
-    public GraknClient(String address) {
-        channel = ManagedChannelBuilder.forTarget(address).usePlaintext().build();
-        databases = new RPCDatabaseManager(channel);
-    }
+        public Core(String address) {
+            channel = ManagedChannelBuilder.forTarget(address).usePlaintext().build();
+            databases = new RPCDatabaseManager.Core(channel);
+        }
 
-    @Override
-    public Session session(String database, Session.Type type) {
-        return session(database, type, new GraknOptions());
-    }
+        @Override
+        public RPCSession.Core session(String database, Grakn.Session.Type type) {
+            return session(database, type, new GraknOptions());
+        }
 
-    @Override
-    public Session session(String database, Session.Type type, GraknOptions options) {
-        return new RPCSession(this, database, type, options);
-    }
+        @Override
+        public RPCSession.Core session(String database, Grakn.Session.Type type, GraknOptions options) {
+            return new RPCSession.Core(this, database, type, options);
+        }
 
-    @Override
-    public DatabaseManager databases() {
-        return databases;
-    }
+        @Override
+        public RPCDatabaseManager.Core databases() {
+            return databases;
+        }
 
-    @Override
-    public boolean isOpen() {
-        return !channel.isShutdown();
-    }
+        @Override
+        public boolean isOpen() {
+            return !channel.isShutdown();
+        }
+
+        @Override
+        public void close() {
+            try {
+                channel.shutdown().awaitTermination(10, TimeUnit.SECONDS);
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+            }
+        }
 
-    @Override
-    public void close() {
-        try {
-            channel.shutdown().awaitTermination(10, TimeUnit.SECONDS);
-        } catch (InterruptedException e) {
-            Thread.currentThread().interrupt();
+        public Channel channel() {
+            return channel;
         }
     }
 
-    public Channel channel() {
-        return channel;
+    public static class Cluster implements Grakn.Client {
+        private static final Logger LOG = LoggerFactory.getLogger(Cluster.class);
+        private final ConcurrentMap<Address.Cluster.Server, Core> coreClientMap;
```

Comment:
Random thought - why is this a `ConcurrentMap` (as opposed to a simple `HashMap`)? I can't see anywhere where we would do parallel writes to it.

Reply:
Good point! Fixed in 1408f4936a972af937084901269da8c46b0adc63